File: solutions/Golang/n_queens_ii_backtracking_optimized.go
package main

import "fmt"

func totalNQueens(n int) int {
    cols := make(map[int]bool)
    diag1 := make(map[int]bool)
    diag2 := make(map[int]bool)
    return backtrack(0, cols, diag1, diag2, n)
}

func backtrack(row int, cols, diag1, diag2 map[int]bool, n int) int {
    if row == n { return 1 }
    count := 0
    for col := 0; col < n; col++ {
        d1, d2 := row-col, row+col
        if !cols[col] && !diag1[d1] && !diag2[d2] {
            cols[col], diag1[d1], diag2[d2] = true, true, true
            count += backtrack(row+1, cols, diag1, diag2, n)
            delete(cols, col)
            delete(diag1, d1)
            delete(diag2, d2)
        }
    }
    return count
}

File: solutions/JavaScript/n_queens_ii_backtracking_optimized.js
function totalNQueens(n) {
    function backtrack(row, cols, diag1, diag2) {
        if (row === n) return 1;
        let count = 0;
        for (let col = 0; col < n; col++) {
            const d1 = row - col, d2 = row + col;
            if (!cols.has(col) && !diag1.has(d1) && !diag2.has(d2)) {
                cols.add(col);
                diag1.add(d1);
                diag2.add(d2);
                count += backtrack(row + 1, cols, diag1, diag2);
                cols.delete(col);
                diag1.delete(d1);
                diag2.delete(d2);
            }
        }
        return count;
    }
    return backtrack(0, new Set(), new Set(), new Set());
}
